File: src/cpp/src/utils/no_metals/preparation.cpp
#include "molgr/utils/no_metals/preparation.h"

#include "molgr/stages/clean.h"
#include "molgr/stages/eliminate.h"
#include "molgr/stages/fresh.h"
#include "molgr/stages/preprocess.h"
#include "molgr/utils/conversions.h"
#include "molgr/utils/electrons.h"
#include "molgr/utils/xyz.h"

#include "molgr/compat/openbabel_iter.h"

#include <memory>
#include <utility>

namespace
{
    // Electron bookkeeping: discard XYZ/Open Babel inferred formal charges and
    // all three MolGR electron classifications before rebuilding from topology
    // and the supplied global budgets.
    void NormalizeSeedElectronicLabels(OpenBabel::OBMol &mol)
    {
        FOR_ATOMS_OF_MOL(atom_iter, mol)
        {
            atom_iter->SetFormalCharge(0);
            molgr::utils::SetUnpairedElectronCount(*atom_iter, 0);
            molgr::utils::SetLonePairCount(*atom_iter, 0);
            molgr::utils::SetUnresolvedTwoElectronCenter(*atom_iter, false);
        }
    }
}

namespace molgr
{
    namespace no_metals
    {
        namespace preparation
        {
            molgr::state::ReconstructionState BuildSeedState(
                std::shared_ptr<OpenBabel::OBMol> omol,
                int total_charge,
                int total_radical_electrons)
            {
                return molgr::state::ReconstructionState(
                    std::move(omol),
                    0,
                    total_charge,
                    total_radical_electrons,
                    {"read_xyz", "normalize_seed_electronic_labels"},
                    {{"source", std::string("xyz_to_omol_no_metal_state")}},
                    0);
            }

            std::shared_ptr<OpenBabel::OBMol> SeedOmolFromXyzBlock(
                const std::string &xyz_block)
            {
                auto omol = std::make_shared<OpenBabel::OBMol>();
                if (!molgr::utils::ReadXyzBlockToMol(xyz_block, omol.get()))
                {
                    return nullptr;
                }
                NormalizeSeedElectronicLabels(*omol);
                return omol;
            }

            std::shared_ptr<OpenBabel::OBMol> NormalizeSeedOmolCopy(
                const OpenBabel::OBMol &seed_omol)
            {
                auto omol = std::make_shared<OpenBabel::OBMol>(
                    molgr::utils::CloneMolTopologyOnly(seed_omol));
                NormalizeSeedElectronicLabels(*omol);
                return omol;
            }

            molgr::state::ReconstructionState SeedStateFromOmol(
                const OpenBabel::OBMol &seed_omol,
                int total_charge,
                int total_radical_electrons)
            {
                return BuildSeedState(
                    NormalizeSeedOmolCopy(seed_omol),
                    total_charge,
                    total_radical_electrons);
            }

            molgr::state::ReconstructionState SeedState(
                const std::string &xyz_block,
                int total_charge,
                int total_radical_electrons)
            {
                auto omol = SeedOmolFromXyzBlock(xyz_block);
                if (!omol)
                {
                    return {};
                }
                return BuildSeedState(
                    std::move(omol),
                    total_charge,
                    total_radical_electrons);
            }

            molgr::state::ReconstructionState PrepareNoMetalSeed(
                const molgr::state::ReconstructionState &state)
            {
                auto machine = molgr::state::OmolStateMachine::FromReconstructionState(state);
                machine.RunOmolStage("make_connections", reconstruct::MakeConnections, 0.15);
                machine.RunOmolStage("pre_clean", reconstruct::PreClean);
                machine.RunOmolStage(
                    "fresh_omol_charge_radical_initial",
                    reconstruct::FreshOmolChargeRadical);

                int formal_charge_sum = 0;
                FOR_ATOMS_OF_MOL(atom_iter, machine.EnsureUniqueMol())
                {
                    formal_charge_sum += atom_iter->GetFormalCharge();
                }
                machine.SetGivenCharge(
                    "initialize_charge_budget",
                    state.total_charge - formal_charge_sum);
                machine.RunOmolChargeStage(
                    "eliminate_NNN_negative",
                    reconstruct::EliminateNNN,
                    false);
                machine.RunOmolChargeStage(
                    "eliminate_high_positive_charge_atoms",
                    reconstruct::EliminateHighPositiveChargeAtoms);
                machine.RunOmolChargeStage(
                    "eliminate_CN_in_doubt",
                    reconstruct::EliminateCNInDoubt);
                machine.RunOmolChargeStage(
                    "eliminate_NNN_positive",
                    reconstruct::EliminateNNN,
                    true);
                machine.RunOmolChargeStage(
                    "eliminate_carboxyl",
                    reconstruct::EliminateCarboxyl);
                machine.RunOmolStage(
                    "clean_carbene_neighbor_unsaturated_first",
                    reconstruct::CleanCarbeneNeighborUnsaturated);
                machine.RunOmolChargeStage(
                    "eliminate_carbene_neighbor_heteroatom",
                    reconstruct::EliminateCarbeneNeighborHeteroatom);
                machine.Annotate("prepare_no_metal_seed");
                return machine.FreezeLike(state);
            }
        }
    }
}

File: src/molgr/fallback/utils/no_metals/preparation.py
"""Preparation helpers for no-metal fallback reconstruction."""

from __future__ import annotations

from typing import cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.fallback.state import ReconstructionState
from molgr.fallback.utils.electrons import (
    set_lone_pair_count,
    set_unpaired_electron_count,
    set_unresolved_two_electron_center,
)


def _normalize_seed_electronic_labels(omol: pybel.Molecule) -> pybel.Molecule:
    """Clear all inferred seed charges and explicit MolGR electron classifications.

    XYZ/Open Babel perception is not trusted as an electronic-state source. The
    no-metal pipeline therefore resets formal charge, real unpaired electrons,
    active lone pairs, and unresolved-center markers before rebuilding them from
    topology and the supplied global budgets.
    """

    for atom in omol.atoms:
        atom.OBAtom.SetFormalCharge(0)
        set_unpaired_electron_count(atom.OBAtom, 0)
        set_lone_pair_count(atom.OBAtom, 0)
        set_unresolved_two_electron_center(atom.OBAtom, False)
    return omol


def _seed_state(
    xyz_block: str,
    total_charge: int,
    total_radical_electrons: int,
) -> ReconstructionState:
    """Create the initial reconstruction state from the input XYZ block."""

    return _seed_state_from_omol(
        _seed_omol_from_xyz(xyz_block),
        total_charge,
        total_radical_electrons,
    )


def _seed_omol_from_xyz(xyz_block: str) -> pybel.Molecule:
    """Parse and normalize the shared initial no-metal molecule from XYZ."""

    return _normalize_seed_electronic_labels(pybel.readstring("xyz", xyz_block))


def _seed_state_from_omol(
    seed_omol: pybel.Molecule,
    total_charge: int,
    total_radical_electrons: int,
) -> ReconstructionState:
    """Create an independent initial state by cloning a normalized seed molecule."""

    omol = _normalize_seed_electronic_labels(
        pybel.Molecule(ob.OBMol(cast(ob.OBMol, seed_omol.OBMol)))
    )
    return ReconstructionState(
        omol=omol,
        given_charge=0,
        total_charge=total_charge,
        total_radical_electrons=total_radical_electrons,
        phase_history=("read_xyz", "normalize_seed_electronic_labels"),
        metadata={"source": "xyz_to_omol_no_metal_state"},
    )


def prepare_no_metal_seed(state: ReconstructionState) -> ReconstructionState:
    """Compatibility entry point for the pipeline-owned linear reconstruction."""

    from molgr.fallback.pipeline.reconstruct_without_metals import _run_linear_preparation

    return _run_linear_preparation(state)


__all__ = [
    "_seed_omol_from_xyz",
    "_seed_state",
    "_seed_state_from_omol",
    "prepare_no_metal_seed",
]
